// decentraland-ad/lib/Ad/createContent/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Locale;
(function (Locale) {
    Locale["en"] = "en";
    Locale["es"] = "es";
    Locale["fr"] = "fr";
    Locale["ja"] = "ja";
    Locale["zh"] = "zh";
    Locale["ko"] = "ko";
})(Locale = exports.Locale || (exports.Locale = {}));
var ContentType;
(function (ContentType) {
    // interactive component
    ContentType["Image"] = "image";
    ContentType["Button"] = "button";
    ContentType["Action"] = "action";
    ContentType["TextBlock"] = "textBlock";
    // translation component
    // Translation = 'translation',
    // text component
    ContentType["Link"] = "link";
    ContentType["Title"] = "title";
    ContentType["Subtitle"] = "subtitle";
    ContentType["Paragraph"] = "paragraph";
    ContentType["Strong"] = "strong";
    ContentType["Underline"] = "underline";
    ContentType["Emphasis"] = "emphasis";
    ContentType["Delete"] = "delete";
})(ContentType = exports.ContentType || (exports.ContentType = {}));
