// decentraland-ui/dist/components/SignIn/SignIn.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignIn = void 0;
var React = __importStar(require("react"));
var Button_1 = require("../Button/Button");
var Header_1 = require("../Header/Header");
var StarWalletIcon_1 = require("../StarWalletIcon/StarWalletIcon");
require("./SignIn.css");
var SignIn = /** @class */ (function (_super) {
    __extends(SignIn, _super);
    function SignIn() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SignIn.prototype.render = function () {
        var _a = this.props, center = _a.center, isConnecting = _a.isConnecting, isConnected = _a.isConnected, onConnect = _a.onConnect, hasError = _a.hasError, i18n = _a.i18n;
        var classes = 'SignIn';
        if (center) {
            classes += ' center';
        }
        var errorClasses = 'error';
        if (hasError && !isConnecting && !isConnected) {
            errorClasses += ' visible';
        }
        return (React.createElement("div", { className: classes },
            React.createElement(Header_1.Header, null, i18n.header),
            React.createElement(StarWalletIcon_1.StarWalletIcon, null),
            React.createElement("p", { className: "message" }, i18n.message),
            React.createElement(Button_1.Button, { primary: true, onClick: onConnect, disabled: isConnecting || isConnected }, isConnecting
                ? i18n.connecting
                : isConnected
                    ? i18n.connected
                    : i18n.connect),
            React.createElement("p", { className: errorClasses }, i18n.error)));
    };
    SignIn.defaultProps = {
        center: true,
        isConnected: false,
        isConnecting: false,
        hasError: false,
        onConnect: function () { return undefined; },
        i18n: {
            header: 'Get Started',
            message: (React.createElement("span", null,
                "You can use the",
                ' ',
                React.createElement("a", { href: "https://metamask.io", target: "_blank", rel: "noopener noreferrer" }, "MetaMask"),
                ' ',
                "extension or a hardware wallet like",
                ' ',
                React.createElement("a", { href: "https://www.ledger.com/", target: "_blank", rel: "noopener noreferrer" }, "Ledger Nano S"),
                ".")),
            connect: 'Connect',
            connecting: 'Connecting...',
            connected: 'Connected',
            error: 'Could not connect to wallet.'
        }
    };
    return SignIn;
}(React.PureComponent));
exports.SignIn = SignIn;
