// decentraland-ui/dist/components/LanguageDropdown/LanguageDropdown.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageDropdown = void 0;
var React = __importStar(require("react"));
var Dropdown_1 = require("../Dropdown/Dropdown");
var LanguageIcon_1 = require("../LanguageIcon/LanguageIcon");
require("./LanguageDropdown.css");
var LanguageDropdown = /** @class */ (function (_super) {
    __extends(LanguageDropdown, _super);
    function LanguageDropdown() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.renderOption = function (locale) {
            var i18n = _this.props.i18n;
            return React.createElement(LanguageIcon_1.LanguageIcon, { locale: locale, label: i18n[locale] });
        };
        return _this;
    }
    LanguageDropdown.prototype.render = function () {
        var _this = this;
        var _a = this.props, locale = _a.locale, locales = _a.locales, direction = _a.direction, upward = _a.upward, onChange = _a.onChange;
        return (React.createElement(Dropdown_1.Dropdown, { className: "dcl language-dropdown", direction: direction, upward: upward, defaultValue: locale == null ? 'en' : undefined, value: locale, options: locales.map(function (key) { return ({
                key: key,
                value: key,
                text: _this.renderOption(key)
            }); }), onChange: onChange }));
    };
    LanguageDropdown.defaultProps = {
        onChange: function () { return undefined; },
        direction: 'left',
        upward: false,
        locales: ['en'],
        i18n: {
            en: 'English',
            es: 'Spanish',
            fr: 'French',
            ja: 'Japanese',
            zh: 'Chinese',
            ko: 'Korean'
        }
    };
    return LanguageDropdown;
}(React.PureComponent));
exports.LanguageDropdown = LanguageDropdown;
