// @dcl/builder-client/dist/files/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WearableConfigSchema = exports.BuilderConfigSchema = void 0;
const schemas_1 = require("@dcl/schemas");
const types_1 = require("../item/types");
exports.BuilderConfigSchema = {
    type: 'object',
    properties: {
        id: {
            type: 'string',
            nullable: true
        },
        collectionId: {
            type: 'string',
            nullable: true
        }
    },
    additionalProperties: false,
    required: []
};
exports.WearableConfigSchema = {
    type: 'object',
    properties: {
        id: {
            type: 'string',
            nullable: true
        },
        description: {
            type: 'string',
            nullable: true
        },
        rarity: {
            ...types_1.Rarity.schema,
            nullable: true
        },
        name: {
            type: 'string'
        },
        data: {
            type: 'object',
            properties: {
                replaces: {
                    type: 'array',
                    items: types_1.WearableCategory.schema
                },
                hides: {
                    type: 'array',
                    items: types_1.WearableCategory.schema
                },
                tags: {
                    type: 'array',
                    items: {
                        type: 'string',
                        minLength: 1
                    }
                },
                representations: {
                    type: 'array',
                    items: schemas_1.WearableRepresentation.schema,
                    minItems: 1
                },
                category: types_1.WearableCategory.schema
            },
            required: ['replaces', 'hides', 'tags', 'representations', 'category']
        }
    },
    additionalProperties: false,
    required: ['name', 'data']
};
